### ISSUE TITLE
FTBFS errors building AF 3.3.alpha for selected architectures

### ISSUE THREAD

**Issue by ghisvail** [arrayfire/arrayfire#1271]
As promised, I ran the pre-release tagged version on the Debian builders. A few architectures failed to build from source:

- [armel](https://buildd.debian.org/status/fetch.php?pkg=arrayfire&arch=armel&ver=3.3~alpha%2Bdfsg1-1&stamp=1454987821): build failure

- [s390x](https://buildd.debian.org/status/fetch.php?pkg=arrayfire&arch=s390x&ver=3.3%7Ealpha%2Bdfsg1-1&stamp=1454978896): test failure

- [hppa](https://buildd.debian.org/status/fetch.php?pkg=arrayfire&arch=hppa&ver=3.3~alpha%2Bdfsg1-1&stamp=1454975180): build failure

- [sparc64](https://buildd.debian.org/status/fetch.php?pkg=arrayfire&arch=sparc64&ver=3.3~alpha%2Bdfsg1-1&stamp=1454975254): build failure

- [x32](https://buildd.debian.org/status/fetch.php?pkg=arrayfire&arch=x32&ver=3.3~alpha%2Bdfsg1-1&stamp=1454975961): build failure

The summary is available [here](https://buildd.debian.org/status/package.php?p=arrayfire&suite=experimental). FYI, amongst this list, only `armel` and `s390x` are considered official platforms on Debian, so we should probably start with these.

**Comment by 9prady9** [comment]
Updating with details for future reference.

s390x
```
[ RUN      ] Asssign.LinearAssignSeq
/«BUILDDIR»/arrayfire-3.3~alpha+dfsg1/test/assign.cpp:886: Failure
Value of: af_assign_seq(&out_arr, in_arr, 1, &ii.idx.seq, rhs_arr)
  Actual: 204
Expected: AF_SUCCESS
Which is: 0
```
```
[  FAILED  ] ImageIO.LoadImage16CPP
[  FAILED  ] ImageMem.SaveMemBMP
[  FAILED  ] ImageIONative.LoadImageNative16SmallCPP
[  FAILED  ] ImageIONative.LoadImageNative16ColorCPP
[  FAILED  ] ImageIONative.LoadImageNative16GrayCPP
```
ImageIO tests failed with value errors and one with the following error
```
[ RUN      ] ImageMem.SaveMemBMP
unknown file: Failure
C++ exception with description "ArrayFire Exception (Function not supported:301):
In function af_err af_load_image_memory(void**, const void*)
In file src/api/c/imageio.cpp:448
FreeImage Error: Unknown File or Filetype

In function af::array af::loadImageMem(const void*)
In file src/api/cpp/imageio.cpp:28" thrown in the test body.
[  FAILED  ] ImageMem.SaveMemBMP (3 ms)
```

**Comment by 9prady9** [comment]
armel

```
In file included from /«BUILDDIR»/arrayfire-3.3~alpha+dfsg1/src/backend/cpu/queue.hpp:11:0,
                 from /«BUILDDIR»/arrayfire-3.3~alpha+dfsg1/src/backend/cpu/Array.hpp:24,
                 from /«BUILDDIR»/arrayfire-3.3~alpha+dfsg1/src/backend/cpu/where.cpp:14:
/«BUILDDIR»/arrayfire-3.3~alpha+dfsg1/src/backend/cpu/threads/async_queue.hpp: In member function 'void async_queue::sync()':
/«BUILDDIR»/arrayfire-3.3~alpha+dfsg1/src/backend/cpu/threads/async_queue.hpp:88:28: error: aggregate 'std::promise<void> p' has incomplete type and cannot be defined
         std::promise<void> p;
                            ^
/«BUILDDIR»/arrayfire-3.3~alpha+dfsg1/src/backend/cpu/threads/async_queue.hpp:89:27: error: variable 'std::future<void> fut' has initializer but incomplete type
         std::future<void> fut = p.get_future();
                           ^
/«BUILDDIR»/arrayfire-3.3~alpha+dfsg1/src/backend/cpu/threads/async_queue.hpp: In lambda function:
/«BUILDDIR»/arrayfire-3.3~alpha+dfsg1/src/backend/cpu/threads/async_queue.hpp:91:15: error: invalid use of incomplete type 'class std::promise<void>'
             pr.set_value();
               ^
In file included from /«BUILDDIR»/arrayfire-3.3~alpha+dfsg1/src/backend/cpu/threads/async_queue.hpp:9:0,
                 from /«BUILDDIR»/arrayfire-3.3~alpha+dfsg1/src/backend/cpu/queue.hpp:11,
                 from /«BUILDDIR»/arrayfire-3.3~alpha+dfsg1/src/backend/cpu/Array.hpp:24,
                 from /«BUILDDIR»/arrayfire-3.3~alpha+dfsg1/src/backend/cpu/where.cpp:14:
/usr/include/c++/5/future:124:11: note: declaration of 'class std::promise<void>'
     class promise;
           ^
In file included from /«BUILDDIR»/arrayfire-3.3~alpha+dfsg1/src/backend/cpu/queue.hpp:11:0,
                 from /«BUILDDIR»/arrayfire-3.3~alpha+dfsg1/src/backend/cpu/Array.hpp:24,
                 from /«BUILDDIR»/arrayfire-3.3~alpha+dfsg1/src/backend/cpu/where.cpp:14:
/«BUILDDIR»/arrayfire-3.3~alpha+dfsg1/src/backend/cpu/threads/async_queue.hpp: In member function 'void async_queue::sync()':
/«BUILDDIR»/arrayfire-3.3~alpha+dfsg1/src/backend/cpu/threads/async_queue.hpp:93:31: error: 'ref' was not declared in this scope
         this->enqueue(f, ref(p));
                               ^
/«BUILDDIR»/arrayfire-3.3~alpha+dfsg1/src/backend/cpu/threads/async_queue.hpp:93:31: note: suggested alternative:
In file included from /usr/include/c++/5/memory:79:0,
                 from /«BUILDDIR»/arrayfire-3.3~alpha+dfsg1/src/backend/cpu/TNJ/Node.hpp:14,
                 from /«BUILDDIR»/arrayfire-3.3~alpha+dfsg1/src/backend/cpu/Array.hpp:18,
                 from /«BUILDDIR»/arrayfire-3.3~alpha+dfsg1/src/backend/cpu/where.cpp:14:
/usr/include/c++/5/functional:454:5: note:   'std::ref'
     ref(reference_wrapper<_Tp> __t) noexcept
```

**Comment by 9prady9** [comment]
`sparc64` and `x32` has same error with GCC 6.0.0

```
cd "/«BUILDDIR»/arrayfire-3.3~alpha+dfsg1/obj-x86_64-linux-gnux32/src/backend/cpu" && /usr/bin/c++   -DAF_CPU -DOS_LNX -DWITH_CPU_LINEAR_ALGEBRA -DWITH_FREEIMAGE -Dafcpu_EXPORTS -I"/«BUILDDIR»/arrayfire-3.3~alpha+dfsg1/include" -I"/«BUILDDIR»/arrayfire-3.3~alpha+dfsg1/src/backend" -I"/«BUILDDIR»/arrayfire-3.3~alpha+dfsg1/src/api/c" -I"/«BUILDDIR»/arrayfire-3.3~alpha+dfsg1/src/backend/cpu" -I"/«BUILDDIR»/arrayfire-3.3~alpha+dfsg1/src/backend/cpu/threads"  -g -O2 -fPIE -fstack-protector-strong -Wformat -Werror=format-security -Wall -pedantic -Wdate-time -D_FORTIFY_SOURCE=2  -fPIC -fPIC   -Wall -std=c++11 -fvisibility=hidden -o CMakeFiles/afcpu.dir/histogram.cpp.o -c "/«BUILDDIR»/arrayfire-3.3~alpha+dfsg1/src/backend/cpu/histogram.cpp"
In file included from /usr/include/c++/6/functional:55:0,
                 from /usr/include/c++/6/memory:79,
                 from /«BUILDDIR»/arrayfire-3.3~alpha+dfsg1/src/backend/cpu/TNJ/Node.hpp:14,
                 from /«BUILDDIR»/arrayfire-3.3~alpha+dfsg1/src/backend/cpu/Array.hpp:18,
                 from /«BUILDDIR»/arrayfire-3.3~alpha+dfsg1/src/backend/cpu/histogram.cpp:13:
/usr/include/c++/6/tuple: In static member function 'static bool std::_Function_base::_Base_manager<_Functor>::_M_manager(std::_Any_data&, const std::_Any_data&, std::_Manager_operation) [with _Functor = std::_Bind<void (*(cpu::Array<unsigned int>, cpu::Array<unsigned int>, unsigned int, double, double))(cpu::Array<unsigned int>, cpu::Array<unsigned int>, unsigned int, double, double)>]':
/usr/include/c++/6/tuple:215:17: internal compiler error: in assign_temp, at function.c:961
       constexpr _Tuple_impl(const _Tuple_impl&) = default;
```

**Comment by 9prady9** [comment]
hppa again with GCC 6.0.0

```
In file included from /<<BUILDDIR>>/arrayfire-3.3~alpha+dfsg1/src/backend/cpu/random.cpp:17:0:
/<<BUILDDIR>>/arrayfire-3.3~alpha+dfsg1/src/backend/cpu/kernel/random.hpp: In function 'void cpu::kernel::randn(cpu::Array<T>) [with T = float]':
/<<BUILDDIR>>/arrayfire-3.3~alpha+dfsg1/src/backend/cpu/kernel/random.hpp:126:9: internal compiler error: Segmentation fault
         my_seed = getSeed();
         ^~~~~~~

0x7177db crash_signal
```

**Comment by pavanky** [comment]
Oh boy. This is going to be rough.

**Comment by pavanky** [comment]
@ghisvail Do you know if we can detect `armel` and `s390x` from either cmake or gcc (using compiler flags perhaps) ?

The armel failure seems to occur because of missing hardware support for some C++11 features. We could disable that part of arrayfire for that hardware using a simple patch.

Not entirely sure why 390x is failing, but I want to try to disable async threads on that architecture as well and see if that solves the issue.

**Comment by pavanky** [comment]
nvm, figured out we need to use `_M_ARCH >= 6` to fix ARMEL issue.

**Comment by pavanky** [comment]
Hmm that seems to be a VS specific flag..

**Comment by ghisvail** [comment]
@pavanky  All the architecture-specific macros can be found [here](https://sourceforge.net/p/predef/wiki/Architectures/).

> The armel failure seems to occur because of missing hardware support for some C++11 features. We could disable that part of arrayfire for that hardware using a simple patch.

If you know which one(s) in particular, I can test providing the corresponding inhibition flag in the rules file. Unless you already have a patch available which I can cheery-pick :-)

**Comment by pavanky** [comment]
@ghisvail We tried to fix a bunch of issues as best as we could. There is no way we can test this properly before 3.3.0 release. So I'm changing this to be 3.3.1 mile stone.

As far as I can tell 3.3.0 should work with armel, but I am not sure about S390x.

**Comment by ghisvail** [comment]
No problem. I can still submit this to the experimental to get feedback from the builders anyway.

**Comment by pavanky** [comment]
@ghisvail Now that 3.3.0 is released can you test again ? Right now we suggest you turn off CPU ASYNC because there may be intermittent bugs. You can do this by adding the following flag to cmake `-DBUILD_CPU_ASYNC=OFF`

**Comment by ghisvail** [comment]
For all architectures ?

**Comment by pavanky** [comment]
@ghisvail x86 and Armv7 more or less works for our tests (for gcc > 4.7.2). It's your call at this point.

**Comment by ghisvail** [comment]
Alright, I'll test without first and provide an update for the architectures where async is confirmed working. Congrats to the team for the release btw.

**Comment by pavanky** [comment]
> Congrats to the team for the release btw.

Thanks :) but this was a smaller release than we'd hoped for.

**Comment by pavanky** [comment]
@ghisvail can you check the build status with `3.3.1` ?

**Comment by pavanky** [comment]
@ghisvail I think it is safe to say we don't have the resources to debug issues on platforms we do not have access to.

**Comment by ghisvail** [comment]
I ended up disabling the `imageio` tests which made all builds succeed. All major architectues build cleanly [now](https://buildd.debian.org/status/logs.php?pkg=arrayfire&ver=3.3.2%2Bdfsg1-4). Perhaps, it would be worth closing this bug now as `WONTFIX`.

Still  a bit surprised that `amd64` succeeds but not `kfreebsd-amd64` due to a test failure with `index_cpu`. The same failure also happened during a refresh of our CI [testing](https://ci.debian.net/data/packages/unstable/amd64/a/arrayfire/20160908_161206.autopkgtest.log.gz) on `amd64`. Perhaps this is already fixed in 3.4 though?

**Comment by pavanky** [comment]
@ghisvail that is indeed weird.. I am not entirely sure why that might happen.

We should be able to support `amd64` for other kernels / operating systems (albeit a bit slowly) as it is mostly a matter of getting a vm up and running.

If you see the same problem happening again on 3.4, please let us know.
